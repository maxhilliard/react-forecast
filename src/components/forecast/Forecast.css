.forecastContainer {
    composes: forecastContainer from '../App.css';

    display: grid;
    grid-template-columns: 100%;
    grid-template-rows: repeat(5, 1fr);
    align-content: center;
    align-items: center;

    border: 4px solid white;

}

.verticalCenter {
    grid-template-rows: 1fr;
}