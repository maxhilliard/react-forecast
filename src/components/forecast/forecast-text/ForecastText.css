.forecastText {
    padding: 20px;
}