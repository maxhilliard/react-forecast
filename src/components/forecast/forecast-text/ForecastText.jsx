import React from 'react';
import PropTypes from 'prop-types';

import styles from './ForecastText.css';

const TEXT_MAP = {
    INSTRUCTIONS: 'Search for a city',
    ERROR: '(◡﹏◡) There was an error with your search',
};

const ForecastText = ({ isForecastError }) => (
    <div className={styles.forecastText}>
        <h2>
            {isForecastError ? TEXT_MAP.ERROR : TEXT_MAP.INSTRUCTIONS}
        </h2>
    </div>
);

ForecastText.propTypes = {
    isForecastError: PropTypes.bool,
};

ForecastText.defaultProps = {
    isForecastError: false,
};

export default ForecastText;
