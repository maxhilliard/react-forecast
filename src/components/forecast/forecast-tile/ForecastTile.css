.forecastTile {
    display: grid;
    grid-template-columns: 33% 33% 33%;
    grid-template-rows: 50% 50%;
    grid-template-areas:
        "forecastDay forecastIcon forecastTemps"
        "forecastDay forecastIcon forecastHumidity";
}

.forecastDay {
    grid-area: forecastDay;
}

.forecastIcon {
    grid-area: forecastIcon;
}

.forecastTemps {
    grid-area: forecastTemps;
}

.forecastHumidity {
    grid-area: forecastHumidity;
}