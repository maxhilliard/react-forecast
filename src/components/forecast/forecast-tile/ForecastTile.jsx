import React from 'react';
import PropTypes from 'prop-types';
import { format } from 'date-fns';

import ForecastDay from './forecast-day/ForecastDay';
import ForecastIcon from './forecast-icon/ForecastIcon';
import ForecastTemps from './forecast-temps/ForecastTemps';
import ForecastHumidity from './forecast-humidity/ForecastHumidity';

import styles from './ForecastTile.css';

// TODO - move this parsing into util
const parseDayFromUnixTimestamp = timestamp => format(timestamp * 1000, 'ddd');

const ForecastTile = ({
    dt,
    humidity,
    temp,
    weather,
}) => {
    const day = parseDayFromUnixTimestamp(dt);
    const { icon, description } = weather[0];

    return (
        <div className={styles.forecastTile}>
            <ForecastDay day={day} />
            <ForecastIcon icon={icon} description={description} />
            <ForecastTemps temp={temp} />
            <ForecastHumidity humidity={humidity} />
        </div>
    );
};

ForecastTile.propTypes = {
    dt: PropTypes.number.isRequired,
    humidity: PropTypes.number.isRequired,
    temp: PropTypes.objectOf(PropTypes.number).isRequired,
    weather: PropTypes.array.isRequired, // eslint-disable-line
};

export default ForecastTile;
