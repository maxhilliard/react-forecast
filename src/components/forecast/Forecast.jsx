import React from 'react';
import PropTypes from 'prop-types';

import ForecastTile from './forecast-tile/ForecastTile';
import ForecastText from './forecast-text/ForecastText';

import styles from './Forecast.css';

const selectRelevantForecastKeys = ({
    dt,
    humidity,
    temp,
    weather,
}) => ({
    dt,
    humidity,
    temp,
    weather,
});

const renderForecastTiles = (allForecastDataForDay) => {
    const forecastTileProps = selectRelevantForecastKeys(allForecastDataForDay);
    return <ForecastTile key={`forecast-${forecastTileProps.dt}`} {...forecastTileProps} />;
};

const Forecast = ({ forecast: fiveDayForecast, isForecastError }) => {
    const shouldRenderText = !fiveDayForecast.length || isForecastError;
    const shouldRenderForecast = !!fiveDayForecast.length && !isForecastError;

    return (
        <div className={`${styles.forecastContainer} ${shouldRenderText ? styles.verticalCenter : ''}`}>
            {shouldRenderText && <ForecastText isForecastError={isForecastError} />}
            {shouldRenderForecast && fiveDayForecast.map(renderForecastTiles)}
        </div>
    );
};

Forecast.propTypes = {
    forecast: PropTypes.array, // eslint-disable-line
    isForecastError: PropTypes.bool,
};

Forecast.defaultProps = {
    forecast: [],
    isForecastError: false,
};

export default Forecast;
