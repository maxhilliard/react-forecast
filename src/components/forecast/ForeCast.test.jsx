import React from 'react';
import { shallow } from 'enzyme';

import Forecast from './Forecast';
import ForecastTile from './forecast-tile/ForecastTile';
import ForecastText from './forecast-text/ForecastText';

describe('Forecast', () => {
    const mockForecast = [
        {
            dt: 'mock-dt',
            humidity: 'mock-humidity',
            temp: 'mock-temp',
            weather: 'mock-weather',
        },
        {
            dt: 'mock-dt',
            humidity: 'mock-humidity',
            temp: 'mock-temp',
            weather: 'mock-weather',
        },
    ];

    const wrapper = shallow(<Forecast forecast={mockForecast} />);

    describe('snapshot', () => {
        it('should match snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });
    });

    describe('props', () => {
        it('should render a ForecastTile component for each item in the forecast array prop', () => {
            const forecastTileCount = wrapper.find(ForecastTile).length;
            const expectedForecastTileCount = mockForecast.length;

            expect(forecastTileCount).toEqual(expectedForecastTileCount);
        });

        it('should render a ForecastText component if there is no forecast', () => {
            wrapper.setProps({ forecast: [] });
            const forecastTextCount = wrapper.find(ForecastText).length;
            const expectedForecastTextCount = 1;

            expect(forecastTextCount).toEqual(expectedForecastTextCount);
        });

        it('should render a ForecastText component if isForecastError prop is truthy', () => {
            wrapper.setProps({ isForecastError: true });
            const forecastTextCount = wrapper.find(ForecastText).length;
            const expectedForecastTextCount = 1;

            expect(forecastTextCount).toEqual(expectedForecastTextCount);
        });
    });
});
