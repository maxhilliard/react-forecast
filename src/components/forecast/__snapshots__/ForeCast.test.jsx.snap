// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forecast snapshot should match snapshot 1`] = `
<div
  className="forecastContainer "
>
  <ForecastTile
    dt="mock-dt"
    humidity="mock-humidity"
    key="forecast-mock-dt"
    temp="mock-temp"
    weather="mock-weather"
  />
  <ForecastTile
    dt="mock-dt"
    humidity="mock-humidity"
    key="forecast-mock-dt"
    temp="mock-temp"
    weather="mock-weather"
  />
</div>
`;
